### DISCUSSION TITLE
How to mock `StreamProvider` for unit testing?

### DISCUSSION THREAD

**Discussion by ggirotto** [rrousselGit/riverpod#3308] [Q&A] [answered]
I have a provider that depends on a `AutoDisposeStreamProvider`. When writing unit test for the first provider, I would like to mock the behavior of my `AutoDisposeStreamProvider` to output a specific value. Here's the code I have so far:

Provider:
```dart
class Provider extends AutoDisposeNotifier<State>  {

  @override
  State build() {
    final result = ref.read(streamProvider);
    return State(result: result);
  }
```

Unit test:
```dart
class ListenerMock<T> extends Mock {
  void call(T? previous, T next);
}

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  late ProviderContainer container;

  final fakeStreamResult = "fake stream result"

  setUp(() {
    container = ProviderContainer(
      overrides: [
        streamProvider.overrideWith((ref) => Stream.value(fakeStreamResult)),
      ],
    );

    addTearDown(container.dispose);
  });

test('unit test', () async {
    final listener = ListenerMock<State>();
    container.listen<State>(provider, listener, fireImmediately: true);

    final expectedState = State(result: fakeStreamResult);

    verifyInOrder(
      [
        () => listener(null, expectedState),
      ],
    );

    verifyNoMoreInteractions(listener);
  });
```

The problem I'm facing is that inside the unit test, when it begins the provider `build` function, the result of doing `ref.read(streamProvider);` returns a `AsyncLoading` instead the mocked `fakeStreamResult` data.

When testing `AsyncNotifierProvider`'s it's possible to do `await container.read(provider.future)` that waits for the `build` function to finish. Do we have something similar to `StreamProvider`?

**Comment by AhmedLSayed9** [comment]
You can use `await container.read(streamProvider.future)` with stream providers.

You also need to use `ref.watch` at the notifier instead of `ref.read` to listen for changes.

**Reply by ggirotto** [reply]
Thanks for the suggestion. When using `await container.read(streamProvider.future)` the same behavior happens: when reading the provider it has `AsyncLoading` state instead the mocked value. Updated test code:

```dart
test('unit test', () async {
    final listener = ListenerMock<State>();
    container.listen<State>(provider, listener, fireImmediately: true);

    await container.read(streamProvider.future);

    final expectedState = State(result: fakeStreamResult);

    verifyInOrder(
      [
        () => listener(null, expectedState),
      ],
    );

    verifyNoMoreInteractions(listener);
  });
```

**Reply by rrousselGit** [reply, marked as answer]
Like mentioned in the test doc, the recommended solution is to not use `container.read` but `container.listen`:

```dart
final sub = container.listen(streamProvider.future, (prev, next) {});
await sub.read();
```

**Reply by ggirotto** [reply]
Thanks for the help!
